applications/IgaApplication/custom_elements/base_discrete_element.h

```diff
@@ -0,0 +1,463 @@
+        std::vector<Matrix>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Set a Constitutive Law Value on the Element
+    * @param rVariable The variable we want to set
+    * @param rValues The values to set in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void SetValueOnIntegrationPoints(
+        const Variable<ConstitutiveLaw::Pointer>& rVariable,
+        std::vector<ConstitutiveLaw::Pointer>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+/********************************************************************/
+/*    GetValueOnIntegrationPoints                                   */
+/********************************************************************/
+    /**
+    * @brief Get on rVariable a array_1d Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<array_1d<double, 3>>& rVariable,
+        std::vector<array_1d<double, 3>>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Get on rVariable a Vector Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<Vector>& rVariable,
+        std::vector<Vector>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Get on rVariable a Matrix Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<Matrix>& rVariable,
+        std::vector<Matrix>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        std::stringstream buffer;
+        buffer << "\"BaseDiscreteElement\" #" << Id();
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "\"BaseDiscreteElement\" #" << Id();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const {
+        pGetGeometry()->PrintData(rOStream);
+    }
+    ///@}
+
+protected:
+
+    ///@name Protected member Variables
+    ///@{
+    std::vector<ConstitutiveLaw::Pointer> mConstitutiveLawVector; /// The vector containing the constitutive laws
```

Comment:
@tteschemacher what do you think?

Reply:
For some element formulations, as thick shells it might be necessary to have more integration points in the thickness - they are referenced to the same spatial position but vary in the thickness.